src/crimson/os/poseidonstore/cache.cc

```diff
@@ -0,0 +1,120 @@
+// -*- mode:C++; tab-width:8; c-basic-offset:2; indent-tabs-mode:t -*-
+// vim: ts=8 sw=2 smarttab
+
+#include "crimson/common/log.h"
+#include "crimson/os/poseidonstore/cache.h"
+#include "crimson/os/poseidonstore/transaction.h"
+#include <vector>
+
+namespace {
+  seastar::logger& logger() {
+    return crimson::get_logger(ceph_subsys_filestore);
+  }
+}
+
+namespace crimson::os::poseidonstore {
+
+std::ostream &operator<<(std::ostream &out, const CachedExtent &ext) {
+  return out << "CachedExtent(type=" << ext.get_type()
+      << ", laddr=" << ext.get_laddr()
+      << ", state=" << ext.state << ")";
+}
+
+std::ostream &operator<<(std::ostream &out, CachedExtent::ce_state_t state) {
+  switch (state) {
+    case CachedExtent::ce_state_t::WRITING:
+      return out << "WRITING";
+    case CachedExtent::ce_state_t::READING:
+      return out << "READING";
+    case CachedExtent::ce_state_t::NEW:
+      return out << "NEW";
+    case CachedExtent::ce_state_t::CLEAN:
+      return out << "CLEAN";
+    default:
+      return out << "UNKNOWN";
+  }
+}
+
+std::optional<RecordRef> Cache::try_construct_record(Transaction &t) 
+{
+  //record_t record;
+  RecordRef record = new Record();
+  for (auto &c : t.write_set) {
+    logger().debug("try_construct_record: add writeset {}", *c);
+    bufferlist bl;
+    // add whole cache extent if modified range is empty
+    if (c->modified_range.empty()) {
+      bl.append(c->get_bptr());
+      assert(bl.length() == c->get_length());
+      record->to_write.push_back(
+	  write_item{
+	    std::move(bl),
+	    c->get_type(),
+	    c->get_laddr()
+	  });
+    } else {
+      bl.append(c->get_bptr(), c->modified_range.range_start(), 
+		    c->modified_range.range_end());
+      record->to_write.push_back(
+	  write_item{
+	    std::move(bl),
+	    c->get_type(),
+	    c->get_laddr()
+	  });
+    }
+  }
```

Comment:
This is a departure from seastore -- currently we assume that "deltas" will be replayed in full from the initial state of the extent.  This variant, however, is something we plan on introducing in the not particularly far future to permit in-place modification for media where that is more expedient.